src/pymor/algorithms/bfgs.py

```diff
@@ -0,0 +1,320 @@
+
+    gradient = model.parameters.parse(model.output_d_mu(mu)).to_numpy()
+    current_output = output(mu)
+    eps = np.linalg.norm(gradient)
+
+    # compute norms
+    mu_norm = np.linalg.norm(mu)
+    update_norms = []
+    foc_norms = []
+    line_search_iterations = []
+    data['mus'] = [mu.copy()]
+
+    hessian = np.eye(mu.size)
+
+    first_order_criticity = output_diff = mu_diff = update_norm = 1e6
+    iteration = 0
+    while True:
+        if iteration >= miniter:
+            if output_diff < rtol:
+                logger.info(f'Relative tolerance of {rtol} for output difference reached. Converged.')
+                break
+            if mu_diff < rtol:
+                logger.info(f'Relative tolerance of {rtol} for parameter difference reached. Converged.')
+                break
+            if first_order_criticity < tol_sub:
+                logger.info(f'Absolute tolerance of {tol_sub} for first order criticity reached. Converged.')
+                break
+            if error_aware:
+                if error_criterion(mu, current_output):
+                    logger.info(f'Output error confidence reached for beta {beta} and radius {radius}. Converged.')
+                    break
+            if (iteration >= stagnation_window + 1 and not stagnation_threshold == np.inf
+                    and all(np.isclose(
+                        [max(update_norms[-stagnation_window - 1:]), min(update_norms[-stagnation_window - 1:])],
+                        update_norm, atol=stagnation_threshold))):
+                logger.info(f'Norm of update is stagnating (threshold: {stagnation_threshold:5e}, '
+                            f'window: {stagnation_window}). Converged.')
+                break
+            if iteration >= maxiter:
+                logger.info(f'Maximum iterations reached. Failed to converge after {iteration} iterations.')
+                raise BFGSError
+
+        iteration += 1
+
+        active, inactive = get_active_and_inactive_sets(parameter_space, mu, eps)
+
+        # compute update to mu
+        if sum(inactive) == 0.:
+            direction = -gradient
+        else:
+            direction = compute_hessian_action_matrix(hessian, -gradient, active, inactive)
+
+        if error_aware:
+            step_size, line_search_iteration = armijo(
+                output, mu, direction, grad=gradient, initial_value=current_output,
+                additional_criterion=error_criterion, **(line_search_params or {}))
+        else:
+            step_size, line_search_iteration = armijo(output, mu, direction, grad=gradient,
+                               initial_value=current_output, **(line_search_params or {}))
+        line_search_iterations.append(line_search_iteration)
+
+        # update mu
+        old_mu = mu.copy()
+        old_output = current_output.copy()
+        mu += step_size * direction
+        mu = parameter_space.clip(mu).to_numpy()
+        current_output = output(mu)
```

Comment:
```suggestion
        current_output = output(mu).item()
```

Reply:
I don't think we actually want this. If I deitemize the value, `.copy()` no longer works because that's a `NumPy` method. Also, this code works fine (for me) without `.item()`, and later generalization to higher dimensional outputs would again break at this point.